### ISSUE TITLE
Port twisted.python.util.InsensitiveDict to Python 3

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6190]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#6190|
|Type|enhancement|
|Created|2012-11-16 20:00:23Z|
|Branch|https://github.com/twisted/twisted/tree/insensitive-py3-6190|

Unfortunately `InsensitiveDict` is required by [#6077](https://github.com/twisted/twisted/issues/6077), so needs to be ported to Python 3.

<details><summary>Searchable metadata</summary>

```
trac-id__6190 6190
type__enhancement enhancement
reporter__itamar itamar
priority__normal normal
milestone__Python_3_3_Minimal Python 3.3 Minimal
branch__branches_insensitive_py3_6190 branches/insensitive-py3-6190
branch_author__itamarst itamarst
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1353096023000000 1353096023000000
changetime__1353424082000000 1353424082000000
version__None None
owner__itamar itamar

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_1"></a>|@itamarst commented|
|-|-|

(In [36416]) Branching to 'insensitive-py3-6190'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_2"></a>|@itamarst set owner to @exarkun|
|-|-|

1.Minor tweaks to make code work on Python 3, and moved to `_utilpy3.py`.
 1. Updated tests so they violate the coding standard less blatantly. They're still terrible.
 1. Added a couple of tests to cover the fact that I had to change `str` to `bytes`.

http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/insensitive-py3-6190

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun set owner to @itamarst|
|-|-|

Thanks.  Just one point - `_doPreserve` does not look completely ported (still using `str` _and_ `unicode`).  Otherwise looks good (I mean, looks terrible, argh, I hate this code, let's get rid of it), please merge after addressing the `_doPreserve` point.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_4"></a>|@itamarst set status to `closed`|
|-|-|

(In [36433]) Merge insensitive-py3-6190.

Author: itamar
Review: exarkun
Fixes: [#6190](https://github.com/twisted/twisted/issues/6190)

Port twisted.python.util.InsensitiveDict to Python 3.
